connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufOptions.scala

```diff
@@ -38,6 +38,12 @@ private[sql] class ProtobufOptions(
 
   val parseMode: ParseMode =
     parameters.get("mode").map(ParseMode.fromString).getOrElse(FailFastMode)
+
+  val circularReferenceType: String = parameters.getOrElse("circularReferenceType", "FIELD_NAME")
```

Comment:
@SandishKumarHN @baganokodo2022 moving the discussion here (for threading).

> Besides, can we also support a "CircularReferenceType" option with a enum value of [FIELD_NAME, FIELD_TYPE]. The reason is because navigation can go very deep before the same fully-qualified FIELD_NAME is encountered again. While FIELD_TYPE stops recursive navigation much faster.  ...

I didn't quite follow the motivation here. Could you give a concrete examples for the two difference cases?

Reply:
@rangadi we already know about field_name circusive check. using `fd.getFullName` we detect the recursion and throw and error.  another option is to detect recursion through field type. example below.

```
message A {
B b;
}

message B {
A c;
}
```
in the case of field_name recursive check it is ```A.B.C```  no recursion. 
in the case of field_type recursive check. it is ```MESSAGE.MESSAGE.MESSAGE``` recursion will be found and throw an error or drop the certain recursive depth. 
but it will also throw an error for the below case with the field_type check. since it will be ```MESSAGE.MESSAGE.MESSAGE.MESSAGE```

```
message A {
B b = 1;
}

message B {
D d = 1;
}

message D {
E e = 1;
}

message E {
int32 key = 1;
}
```
@baganokodo2022 argument is field_type base check will give users an option to drop recursion more quickly because with complex nested schema recursive field_name can be found at very deep. before hitting this we might see OOM. field_type base check finds the circle reference more quickly. 

@baganokodo2022 please correct me if I'm wrong.

Reply:
> in the case of field_name recursive check it is A.B.C no recursion.

That example is clearly recursion. What is 'C' here. 

> but it will also throw an error for the below case with the field_type check. since it will be MESSAGE.MESSAGE.MESSAGE.MESSAGE

Why is this recursion?

Reply:
Are our unit tests showing these cases?

Reply:
I would have @baganokodo2022 give more details on the field type case. 

We have not yet added unit tests for the field-type case. would like to discuss this before adding unit tests.

Reply:
@baganokodo2022 Could you translate that to actual protobuf to illustrate the problem? I still don't understand. 
If the redundat data in the warehouse is concern, customers can process with a smaller protobuf (say with unneeded fields removed).

Reply:
@rangadi , the issue is that protobuf is designed for interactions between micro services. Protobuf schemas are long existing before we introduced Delta lake to the corporate. In many cases, we have to comprise for the needs of micro services, not the other way around. Service developers typically are not motivated or willing to take the risk to revise their data model for data analytics purposes. I will compile a proto schema shortly. cc @SandishKumarHN 

thanks

Reply:
I see only one type of recursion, that is just just based on protobuf message. I.e. 
 message A ends up including itself either directly or indirectly through nesting. This is the 1st one in @SandishKumarHN's message above. The second one does not have recursion. 

Looking for examples of other types.

Reply:
```
message A {
B b = 1;
}
message B {
A aa = 1;
D d = 2;
}

message D {
A aaa = 1;
E a =2;
}

message E {
int32 key = 1;
}
```
@rangadi just correcting the second example. in the case of field_name base recursive check. we would fail to detect the recursion for above because the thread would be ```A.B.A.aa.D.d.A.aaa.E``` set would have unique values [A, B.b, A.aa, A.aaa, E.a], but in the case of @baganokodo2022 proposal it would be ```A.B.A.D.A.E``` would takeout field_name so early detection for recursion.

Reply:
@rangadi just replacing fd.getFullName with below would make sense. we don't need to have two different cases of checks. 
```
val recordName = fd.getFullName.substring(0, fd.getFullName().lastIndexOf("."))
```

Reply:
> we would fail to detect the recursion for above because the thread would be

Why would we fail? Lets say this is doing `from_protobuf(col, 'message_A')`
`A aa = 1` at line 5 would be treated as recursion.
Why are field names relevant at all?

Reply:
Does the code in master fail to detect this recursion? In that case it would lead to unbounded schema and fail.

Reply:
>  thread would be `A.B.A.aa.D.d.A.aaa.E`

What is this thread?

Reply:
Given this discussion, let's write down functionality and examples, before we implement so that we are all on the same page.

Reply:
@rangadi fd.fullName is able to detect the recursive field with different field names. add a unit test. now I'm confused.  
```Fail for recursion field with different field names```

Reply:
:) yeah, field names should not matter at all. 
We can do video chat to clarify all this.

Reply:
@rangadi @baganokodo2022 thanks for the quick meet. meeting conclusion use descriptor type name and added unit tests with some complex schema. 
```
val recordName = fd.getMessageType.getFullName
```